subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/RoutingVirtualFileSystem.java

```diff
@@ -24,22 +24,26 @@
 import org.gradle.internal.vfs.VirtualFileSystem;
 
 import java.util.Optional;
+import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
 public class RoutingVirtualFileSystem implements VirtualFileSystem {
     private final WellKnownFileLocations gradleUserHomeFileLocations;
     private final VirtualFileSystem gradleUserHomeVirtualFileSystem;
     private final VirtualFileSystem buildScopedVirtualFileSystem;
+    private final BooleanSupplier vfsRetained;
 
     public RoutingVirtualFileSystem(
         WellKnownFileLocations gradleUserHomeLocations,
         VirtualFileSystem gradleUserHomeVirtualFileSystem,
-        VirtualFileSystem buildScopedVirtualFileSystem
+        VirtualFileSystem buildScopedVirtualFileSystem,
+        BooleanSupplier vfsRetained
```

Comment:
We could move this class to `:core` and then we could inject something nicer here.